File: LeetCode/remote/insert_interval.cs
// https://leetcode.com/problems/insert-interval/
//
// Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).
//
// You may assume that the intervals were initially sorted according to their start times.
//
// Example 1:
// Given intervals [1,3],[6,9], insert and merge [2,5] in as [1,5],[6,9].
//
// Example 2:
// Given [1,2],[3,5],[6,7],[8,10],[12,16], insert and merge [4,9] in as [1,2],[3,10],[12,16].
//
// This is because the new interval [4,9] overlaps with [3,5],[6,7],[8,10]. 
//
// 
// Submission Details
// 151 / 151 test cases passed.
//  Status: Accepted
//  Runtime: 568 ms
//      
//      Submitted: 0 minutes ago
// 
// 2 errors
// - IList fixed size (probably array)
// - Overlaps needs to be commutative 

using System;
using System.Collections.Generic;
using System.Linq;

public class Interval {
    public int start;
    public int end;
    public Interval() { start = 0; end = 0; }
    public Interval(int s, int e) { start = s; end = e; }
    public override String ToString()
    {
        return String.Format("[{0}, {1}] ", start, end);
    }
}

public class Solution {
    public static IList<Interval> Insert(IList<Interval> intervals, Interval newInterval) {
        intervals = intervals.ToList();
        intervals.Add(newInterval);
        var result = new List<Interval>();
        result.AddRange(intervals.Where(x => x.end < newInterval.start));
        result.Add(intervals
                    .Where(x => Overlaps(x, newInterval) || Overlaps(newInterval, x))
                    .Aggregate(new Interval(int.MaxValue, int.MinValue),
                              (acc, x) => {
                                    acc.start = Math.Min(acc.start, x.start);
                                    acc.end = Math.Max(acc.end, x.end);
                                    return acc;
                              }));
        result.AddRange(intervals.Where(x => x.start > newInterval.end));
        return result;
    }

    static bool Overlaps(Interval a, Interval b)
    {
        return b.start <= a.start && a.start <= b.end ||
               b.start <= a.end && a.end <= b.end;
    }

    static void Main()
    {
        Console.WriteLine(String.Join(", ", Insert(new List<Interval> {
                            new Interval(1, 5)
                        }, new Interval(2, 3))));

        Console.WriteLine(String.Join(", ", Insert(new List<Interval> {
                            new Interval(1, 3),
                            new Interval(6, 9)
                        }, new Interval(2, 5))));

        Console.WriteLine(String.Join(", ", Insert(new List<Interval> {
                            new Interval(1, 2),
                            new Interval(3, 5),
                            new Interval(6, 7),
                            new Interval(8, 10),
                            new Interval(12, 16),
                        }, new Interval(4, 9))));
    }
}

File: LeetCode/remote/insert_interval.rb
#   https://leetcode.com/problems/insert-interval/
#   Given a set of non-overlapping intervals, insert a new interval into the intervals (merge if necessary).
#
#   You may assume that the intervals were initially sorted according to their start times.
#   https://leetcode.com/submissions/detail/54348173/
#   Submission Details
#   151 / 151 test cases passed.
#       Status: Accepted
#       Runtime: 132 ms
#           
#           Submitted: 0 minutes ago

def overlaps(first, second)
    return first.start.between?(second.start, second.end) ||
           first.end.between?(second.start, second.end) ||
           second.start.between?(first.start, first.end) ||
           second.end.between?(first.start, first.end)
end

def merge(first, second)
    Interval.new([first.start, second.start].min, [first.end, second.end].min)
end

def insert(intervals, new_interval)
    intervals << new_interval
    intervals.select { |x| x.end < new_interval.start } +
    [intervals.select { |x| overlaps(x, new_interval) }
             .inject(Interval.new(2 ** (0.size * 8), -(2 ** (0.size * 8)))) { |acc, x|
                Interval.new([acc.start, x.start].min, [acc.end, x.end].max)
             }] +
    intervals.select { |x| x.start > new_interval.end }
end

#   Evolution in a nutshell
#   
#   Submission Details
#   151 / 151 test cases passed.
#       Status: Accepted
#       Runtime: 144 ms
#           
#           Submitted: 0 minutes ago
#   https://leetcode.com/submissions/detail/56512286/
def insert(intervals, new_interval)
    intervals << new_interval
    intervals.select { |x| x.end < new_interval.start } +
    [intervals.select { |x| overlaps?(x, new_interval) }.reduce { |acc, x|
        Interval.new([acc.start, x.start].min, [acc.end, x.end].max)
    }] +
    intervals.select { |x| new_interval.end < x.start}
end

def overlaps?(first, second)
    first.start.between?(second.start, second.end) ||
    first.end.between?(second.start, second.end) ||
    second.start.between?(first.start, first.end) ||
    second.end.between?(first.start, first.end) 
end
